django_mongodb/compiler.py

```diff
@@ -0,0 +1,291 @@
+        """
+        Store a document using field columns as element names.
+
+        If just a {pk_field: None} mapping is given a new empty document is
+        created, otherwise value for a primary key may not be None.
+        """
+        for doc in docs:
+            if doc.get("_id", NOT_PROVIDED) is None:
+                if len(doc) == 1:
+                    # insert with empty model
+                    doc.clear()
+                else:
+                    raise DatabaseError("Can't save entity with _id set to None")
+
+        collection = self.get_collection()
+        options = self.connection.operation_flags.get("save", {})
+
+        if returning_fields:
+            return [collection.insert_one(doc, **options).inserted_id]
+
+        collection.insert_one(doc, **options)
+        return []
+
+
+class SQLDeleteCompiler(SQLCompiler):
+    def execute_sql(self, result_type=MULTI):
+        cursor = Cursor()
+        cursor.rowcount = self.build_query([self.query.get_meta().pk]).delete()
+        return cursor
+
+
+class SQLUpdateCompiler(SQLCompiler):
+    def execute_sql(self, result_type):
+        self.pre_sql_setup()
+        values = []
+        for field, _, value in self.query.values:
+            if hasattr(value, "prepare_database_save"):
+                if field.remote_field:
+                    value = value.prepare_database_save(field)
+                else:
+                    raise TypeError(
+                        f"Tried to update field {field} with a model "
+                        f"instance, {value!r}. Use a value compatible with "
+                        f"{field.__class__.__name__}."
+                    )
+            prepared = field.get_db_prep_save(value, connection=self.connection)
+            values.append((field, prepared))
+        return self.update(values)
+
+    def update(self, values):
+        multi = True
+        spec = {}
+        for field, value in values:
+            if field.primary_key:
+                raise DatabaseError("Cannot modify _id.")
+            if isinstance(value, Expression):
+                raise NotSupportedError("QuerySet.update() with expression not supported.")
+            # .update(foo=123) --> {'$set': {'foo': 123}}
+            spec.setdefault("$set", {})[field.column] = value
+
+            if field.unique:
+                multi = False
+
+        return self.execute_update(spec, multi)
+
+    @safe_call
+    def execute_update(self, update_spec, multi=True, **kwargs):
+        collection = self.get_collection()
+        try:
+            criteria = self.build_query().mongo_query
+        except EmptyResultSet:
+            return 0
+        options = self.connection.operation_flags.get("update", {})
+        options = dict(options, **kwargs)
+        method = "update_many" if multi else "update_one"
+        return getattr(collection, method)(criteria, update_spec, **options).matched_count
+
+
+class SQLAggregateCompiler(SQLCompiler):
```

Comment:
Also, It does not support annotations (`$group` or `group by` in SQL), does it?
If It so, are we going to support it?

Reply:
I think we'll try to support at least some simple aggregations, but I think it'll be complicated. I haven't looked into the details yet.